### DISCUSSION TITLE
Explain how resize works in the background

### DISCUSSION THREAD

**Discussion by ghost** [ImageMagick/ImageMagick#5277] [Help] [answered]
Hi,

How can you visually explain what is actually happening in the background when we run this command?
I want to understand how the -background -resize -gravity -extent fit together

```
mogrify -format png -path ../output -resize "512x512" -background transparent -gravity center -extent 512x512 *.png
```
Thank you

**Comment by snibgo** [comment]
This does two operations.

"-resize 512x512" resizes to the largest size that fits inside a 512x512 box, while retaining the aspect ratio of the input. So the result from that will be exactly 512x512, or 300x512, or 512x234 or whatever.

"-extent 512x512" will add extra pixels to make the result 512x512, if it isn't already. Two setting affect the operation of "-extent". "-background transparent" will make the extra pixels transparent. "-gravity center" will put the input image at the centre of the output, so the extra pixels will be added evenly to the left and right, or top and bottom.

**Reply by ghost** [reply]
@snibgo thank you so much for this detailed explanation, can you please support it with a visual explanation, it will help a lot.
Also does this command do both upscaling and downscaling for the target image?

**Comment by snibgo** [comment, marked as answer]
"-resize 512x512" will shrink or enlarge as required.

Windows syntax:
```
magick ^
  rose: +write r0.png ^
  -resize "512x512" +write r1.png ^
  -background Blue -gravity center -extent 512x512 ^
  r2.png
```
The "rose:" image is 70x46 pixels.

r0.png is 70x46:
![r0](https://user-images.githubusercontent.com/33812368/175823550-bc11c173-83b2-4500-97c9-36506bd6929d.png)

r1.png is 512x336:
![r1](https://user-images.githubusercontent.com/33812368/175823554-317406cb-1855-4c2c-a534-6d3aaaee53f7.png)

r2.png is 512x512:
![r2](https://user-images.githubusercontent.com/33812368/175823585-deca3560-eb38-4191-ab73-d13cbff1b498.png)

In general, inserting "+write x.png" at different places to create debugging images is very useful.

**Reply by ghost** [reply]
Indeed that is a handy feature I did not know exist.
Where do I find those debugging features are they documented somewhere?

Thank you so much

**Comment by snibgo** [comment]
"+write" is documented at https://imagemagick.org/script/command-line-options.php#write . It is a standard IM operation, used whenever we want to write images, so that makes it useful for debugging IM commands. "+write info:" is useful for checking that the image list is as I expect at that point in the command.

**Reply by ghost** [reply]
It is very useful, this is exactly what I was looking for.  l wanted to see what is happening at every step
Thank you for pointing that out
